auto_submit/lib/requests/check_pull_request.dart

```diff
@@ -69,4 +71,24 @@ class CheckPullRequest extends AuthenticatedRequestHandler {
     await Future.wait(futures);
     return Response.ok('Finished processing changes');
   }
+
+  /// Pulls queued Pub/Sub messages.
+  ///
+  /// Pub/Sub pull request API doesn't guarantee returning all messages each time. This
+  /// loops to pull `kPubsubPullNumber` times to try covering all queued messages.
+  Future<List<pub.ReceivedMessage>> pullMessages() async {
+    final Map<String, pub.ReceivedMessage> messageMap = <String, pub.ReceivedMessage>{};
+    for (int i = 0; i < kPubsubPullNumber; i++) {
+      final pub.PullResponse pullResponse = await pubsub.pull('auto-submit-queue-sub', kPullMesssageBatchSize);
+      final List<pub.ReceivedMessage>? receivedMessages = pullResponse.receivedMessages;
+      if (receivedMessages == null) {
+        continue;
+      }
+      for (pub.ReceivedMessage message in receivedMessages) {
+        final String messageId = message.message!.messageId!;
+        messageMap[messageId] = message;
+      }
+    }
+    return messageMap.values.toList();
```

Comment:
Not sure how long it takes for pub/sub api to return but if it takes something more that a couple hundred ms I'd recommend to process the messages as we get the rather than accumulating and processing at the end.

Reply:
What would be the advantage by doing that?
With consolidating first, we just need to iterate all distinct messages once with same goal. But by processing messages following each call, we are repeatedly processing same message 5 times in each cronjob call.
We can keep close monitoring the execution time in the beginning.

Reply:
The main concern is applying synchronous processing logic to asynchronous technologies, ideally in an asynchronous world we get a bunch of messages and delegate the processing to something else without blocking in the processing. There are several subscription configurations to avoid the behavior you mentioned https://cloud.google.com/pubsub/docs/exactly-once-delivery#exactly-once_delivery_guarantees

Reply:
Updated to loop through each` pull` call and process messages immediately in each loop.